✨ [Panning media] Transition siblings by ID

processprocess opened the pull request:
context / fixes #31932

- Adds group-id attribute to associate components, falls back to `amp-img` `src`
- Sets panning media state of active group
- Gets current state before emitting
- Revises state comparison function
- Updates example with parallax effect
- Updates test

[Demo](https://panning-media.web.app/examples/amp-story/amp-story-panning-media.html)

![Jan-27-2021 12-58-34](https://user-images.githubusercontent.com/3860311/106033211-7b527480-609f-11eb-891e-77ef4bbe2532.gif)

mszylkowski commented:
Would it be a good idea to support desktop landscape UI on the demo stories with this feature? Or would it break it?

processprocess commented:
> Would it be a good idea to support desktop landscape UI on the demo stories with this feature? Or would it break it?

`supports-landscape` works as expected: 
![Jan-27-2021 12-52-41](https://user-images.githubusercontent.com/3860311/106032569-a25c7680-609e-11eb-9882-c49a1f6b5fcc.gif)

mszylkowski reviewed (commented):
If you don't specify a group-id then would it sync with the other `null` groups? Or is the API meant to be used with groups even if you don't have multiple sets?

mszylkowski commented on extensions/amp-story-panning-media/0.1/amp-story-panning-media.js:
```diff
@@ -158,19 +164,21 @@ export class AmpStoryPanningMedia extends AMP.BaseElement {
         this.storeService_.dispatch(Action.SET_PANNING_MEDIA_STATE, {
-          x: startPos.x + (this.x_ - startPos.x) * easing,
-          y: startPos.y + (this.y_ - startPos.y) * easing,
-          zoom: startPos.zoom + (this.zoom_ - startPos.zoom) * easing,
+          ...getPanningMediaState(),
+          [this.groupId_]: {
+            x: startPos.x + (this.x_ - startPos.x) * easing,
+            y: startPos.y + (this.y_ - startPos.y) * easing,
+            zoom: startPos.zoom + (this.zoom_ - startPos.zoom) * easing,
+          },
```
You might be able to move this to the `action()` method of `amp-story-store-service.js` so you don't need to fetch the state so many times here and below. You could dispatch `{group: "group-1", data: {x, y, zoom}}` and use the group and data to update the state accordingly, and it could save some calls to `getPanningMediaState`

processprocess replied:
This is awesome. Thank you for providing a detailed sample implementation like this. It helps me understand the full concept more clearly and gives me strong footing to build off of.

I see how this idea is to merge the update on the store level instead of component level.

Do you think there is a benefit to having an additional level with the `group` and `data` keys? 
Otherwise we could do:
`const updatedState = {...state[StateProperty.PANNING_MEDIA_STATE],...data};` 
and keep the map shallow.

mszylkowski replied:
That also works! If I understand you'd still need to key the data with `[data.group]: data`

processprocess commented:
> If you don't specify a group-id then would it sync with the other `null` groups? Or is the API meant to be used with groups even if you don't have multiple sets?

Thanks for pointing this out. I added a fallback for this and updated the example.
if `group-id` is not specified it will fallback to the `amp-img` `src`.
It's likely that if the publisher uses the same image between pages they want it associated so I think that's safe.
`group-id` is then available for creating advanced effects.

Enriqe commented on extensions/amp-story-panning-media/0.1/amp-story-panning-media.js:
```diff
@@ -182,8 +190,7 @@ export class AmpStoryPanningMedia extends AMP.BaseElement {
   onPanningMediaStateChange_(panningMediaState) {
-    if (panningMediaState && this.distance_ <= 1) {
-      // TODO(#31932): Update siblings that are part of the same group.
+    if (panningMediaState[this.groupId_] && this.distance_ <= 1) {
```
sorry I must have missed something but why do we want to animate for neighbor pages?

processprocess replied:
This ensures the frame on the next page is in the correct position. 
When only updating the active page, a noticeable flash of an incorrectly frame is visible:
![Jan-28-2021 12-36-45](https://user-images.githubusercontent.com/3860311/106176652-9687b780-6165-11eb-9fa6-da94b50fb7c7.gif)

Enriqe replied:
Gotcha, thanks for the clarification and gif! super helpful

mszylkowski commented on extensions/amp-story-panning-media/0.1/amp-story-panning-media.js:
```diff
@@ -63,6 +64,12 @@ export class AmpStoryPanningMedia extends AMP.BaseElement {
+    /** @private {?string} */
+    this.groupId_ = null;
+
+    /** @private {!Object<string, string>} Current animation state. */
```
string: float?

mszylkowski reviewed (approved):
One consideration would be that if there are many panning-medias with the same group then it will explode, I wonder if there's a way to validate that

gmajoulet commented on extensions/amp-story-panning-media/0.1/amp-story-panning-media.js:
```diff
@@ -43,14 +44,15 @@ export class AmpStoryPanningMedia extends AMP.BaseElement {
-
-    /** @private {?float} Percentage between [-50; 50] */
-    this.y_ = null;
+    /** @private {?{x: float, y: float, zoom: float}} Position to animate to. */
```
Please create a typedef so you can write better types for animationState_ or animate_()

processprocess commented:
> One consideration would be that if there are many panning-medias with the same group then it will explode, I wonder if there's a way to validate that

Per our conversation, multiple components with the same ID on one page and it still runs smooth without flickering on a low end device.
I think we can communicate this in examples of the documentation.

gmajoulet commented on extensions/amp-story-panning-media/0.1/amp-story-panning-media.js:
```diff
@@ -156,21 +164,22 @@ export class AmpStoryPanningMedia extends AMP.BaseElement {
-          x: startPos.x + (this.x_ - startPos.x) * easing,
-          y: startPos.y + (this.y_ - startPos.y) * easing,
-          zoom: startPos.zoom + (this.zoom_ - startPos.zoom) * easing,
+          [this.groupId_]: {
```
Plz either change the Action to `ADD_*`, or make sure you merge this new data into the old state before setting it, so we get a more accurate action name.

gmajoulet commented on extensions/amp-story/1.0/amp-story-store-service.js:
```diff
@@ -255,12 +255,7 @@ const stateComparisonFunctions = {
-    old.x !== curr.x ||
-    old.y !== curr.y ||
-    old.zoom !== curr.zoom,
+  [StateProperty.PANNING_MEDIA_STATE]: (old, curr) => !deepEquals(old, curr, 2),
```
Can you add an early return if old/curr are empty objects?
